server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -513,6 +513,7 @@ public RetryOnReplicaException(StreamInput in) throws IOException {
         @Override
         public void onResponse(Releasable releasable) {
             try {
+                assert replica.getActiveOperationsCount() != 0 : "must perform shard operation under a permit";
```

Comment:
I think we can even say > 0 here?

Reply:
We can't, because some actions (e.g. `TransportVerifyShardBeforeCloseAction`) run under an operation block rather than a single permit.

Reply:
Makes sense, thanks.